### PR TITLE
[VM] Refactor data_cache/session to prepare for resource groups

### PR DESCRIPTION
**Pull request by gelash**
Isolating some preparatory work from: https://github.com/aptos-labs/aptos-core/pull/9927/

It's similar to https://github.com/aptos-labs/aptos-core/pull/9800, i.e. introduces new gas behavior that we can support in parallel execution, as well as refactors some resolver traits and caches to have less complex changes in 9927

### CODE REVIEW

**Review by igor-aptos — APPROVED** [review]
thanks for adding the tests! it helps documenting the expected logic as well.

Code looks good to me (though I don't have as much context on this code)

can you recheck why is framework_upgrade failing?

**Comment by gelash** [comment]
> thanks for adding the tests! it helps documenting the expected logic as well.
> 
> Code looks good to me (though I don't have as much context on this code)
> 
> can you recheck why is framework_upgrade failing?

indeed - previously I think smt else failed (that seemed more suspicious), but re-running now

**Review by zekun000 — APPROVED** [review]
lead by writing tests!

File: aptos-move/aptos-vm/src/data_cache.rs [hunk 1]
```diff
@@ -107,68 +119,106 @@ impl<'a, S: StateView> StorageAdapter<'a, S> {
...
-                let group_data: BTreeMap<StructTag, Vec<u8>> = bcs::from_bytes(&group_data)
-                    .map_err(|_| {
+
+            let (buf, maybe_group_size) = self
+                .get_resource_from_group(&key, struct_tag, self.accurate_byte_count)
+                .map_err(|e| {
+                    if self.group_byte_count_as_sum {
+                        // Message also gated with the byte counting as sum flag / gas version.
                         PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                            .with_message(format!("{}", e))
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/aptos-vm/src/data_cache.rs:139]: sounds good, let's document here that the info is set to None: https://github.com/aptos-labs/aptos-core/blob/c2153785ea0bacf37a579971e9f38fc26aaeaf2e/types/src/transaction/mod.rs#L869

File: aptos-move/aptos-vm/src/data_cache.rs [hunk 2]
```diff
@@ -107,68 +119,106 @@ impl<'a, S: StateView> StorageAdapter<'a, S> {
...
+        // Resolve directly from state store (StateView interface).
+        let group_data = self.state_store.get_state_value_bytes(key)?;
+        if let Some(group_data_blob) = group_data {
+            let group_data: BTreeMap<StructTag, Vec<u8>> = bcs::from_bytes(&group_data_blob)
+                .map_err(|_| anyhow::Error::msg("Resource group deserialization error"))?;
+
+            let maybe_group_size = if return_group_size {
+                assert!(
+                    self.accurate_byte_count,
+                    "No charge for first access, should not be returning size"
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/aptos-vm/src/data_cache.rs:187]: let's remove the assert, I wanted to make sure it's called only in the context we want, but should be fine anyway

File: aptos-move/aptos-vm/src/data_cache.rs [hunk 3]
```diff
@@ -51,46 +56,53 @@ pub(crate) fn get_resource_group_from_metadata(
 pub struct StorageAdapter<'a, S> {
     state_store: &'a S,
     accurate_byte_count: bool,
+    group_byte_count_as_sum: bool,
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/aptos-vm/src/data_cache.rs:59]: done

**Pull request merged**